components/ScriptLibrary/src/LHVM.cpp

```diff
@@ -106,274 +59,1424 @@ struct imemstream: virtual membuf, std::istream
+}
+
+void LHVM::InvokeStopTaskCallback(const uint32_t taskNumber)
+{
+	if (_stopTaskCallback != nullptr)
+	{
+		_stopTaskCallback(taskNumber);
+	}
+}
+
+void LHVM::AddReference(const uint32_t objectId)
+{
+	if (_addReference != nullptr)
+	{
+		_addReference(objectId);
+	}
+}
+
+void LHVM::RemoveReference(const uint32_t objectId)
+{
+	if (_removeReference != nullptr)
+	{
+		_removeReference(objectId);
+	}
+}
+
+void LHVM::InvokeNativeCallEnterCallback(const uint32_t funcId)
+{
+	if (_nativeCallEnterCallback != nullptr)
+	{
+		_nativeCallEnterCallback(funcId);
 	}
 }
 
-void LHVM::LoadAuto(std::istream& stream)
+void LHVM::InvokeNativeCallExitCallback(const uint32_t funcId)
 {
-	int32_t count;
-	if (!stream.read(reinterpret_cast<char*>(&count), sizeof(count)))
+	if (_nativeCallExitCallback != nullptr)
 	{
-		Fail("error reading id count");
+		_nativeCallExitCallback(funcId);
 	}
+}
 
-	// if there are no variables, return
-	if (count <= 0)
+void LHVM::SignalError(const ErrorCode code)
+{
+	if (_errorCallback != nullptr)
 	{
-		return;
+		_errorCallback(code, "", 0);
 	}
+}
 
-	std::vector<uint32_t> ids(count);
+void LHVM::SignalError(const ErrorCode code, const uint32_t data)
+{
+	if (_errorCallback != nullptr)
+	{
+		_errorCallback(code, "", data);
+	}
+}
 
-	if (!stream.read(reinterpret_cast<char*>(&count), sizeof(ids[0]) * ids.size()))
+void LHVM::SignalError(const ErrorCode code, const std::string& data)
+{
+	if (_errorCallback != nullptr)
 	{
-		Fail("error reading ids");
+		_errorCallback(code, data, 0);
 	}
 }
 
-void LHVM::LoadScripts(std::istream& stream)
+const VMScript* LHVM::GetScript(const std::string& name)
 {
-	int32_t count;
-	if (!stream.read(reinterpret_cast<char*>(&count), sizeof(count)))
+	for (const auto& script : _scripts)
 	{
-		Fail("error reading script count");
+		if (script.GetName() == name)
+		{
+			return &script;
+		}
 	}
+	return nullptr;
+}
 
-	// if there are no variables, return
-	if (count <= 0)
+bool LHVM::TaskExists(const uint32_t taskId)
+{
+	return _tasks.contains(taskId);
+}
+
+uint32_t LHVM::GetTicksCount()
+{
+	return _ticks;
+}
+
+void LHVM::PushElaspedTime()
+{
+	const float time = GetTicksCount() * 10.0f;
+	Pushf(time);
+}
+
+VMVar& LHVM::GetVar(VMTask& task, const uint32_t id)
+{
+	const auto offset = task.variablesOffset;
+	return (id > offset) ? task.localVars.at(id - offset - 1) : _variables.at(id);
+}
+
+uint32_t LHVM::GetExceptionHandlersCount()
+{
+	if (_currentTask != nullptr)
 	{
-		return;
+		return _currentTask->exceptionHandlerIps.size();
 	}
+	return 0;
+}
 
-	char script_name[255];
-	char file_name[255];
+uint32_t LHVM::GetCurrentExceptionHandlerIp(const uint32_t index)
+{
+	if (_currentTask != nullptr)
+	{
+		return _currentTask->exceptionHandlerIps.at(_currentTask->exceptionHandlerIps.size() - index - 1);
+	}
+	return 0;
+}
 
-	for (int32_t i = 0; i < count; i++)
+void LHVM::PrintInstruction(const VMTask& task, const VMInstruction& instruction)
+{
+	// TODO(Daniels118): improve this
```

Comment:
How, why? Maybe make an issue and put details there, then reword as  `// TODO(#ISSUE_NUMBER)`

Reply:
Many opcodes have different meanings depending on both `mode` and `datatype` fields. The current implementation of this print method doesn't take into account all these differences, so different instructions may be shown with identical keywords and we can't really understand what they do (this applies only to this print method, the actual implementation is fully functional). We don't need this right now because the game functions table isn't fully implemented yet, so it would be impossible to run complex CHL scripts, but when it will be ready, we may need correct display of instructions in order to debug the VM and/or the functions implementation.